### ISSUE TITLE
deadlock creating channel on closing connection

### ISSUE THREAD

**Issue by karlseguin** [streadway/amqp#173]
We're seeing cases where opening a channel (`conn.Channel()`) hangs if the connection is in the process of shutting down. Specifically, the `select` in `channel.call` blocks.

It's hard to reproduce reliably, but if you make the following changes to `connection.go` (which helps extend the window in which this can happen), it becomes easy to reproduce.

1 - Make the following 2 changes to connections.go: https://github.com/2nd/amqp/commit/30a656f7430ce5ee6fa85c1efb52633741fe015f

2 - Run the following code

``` go
func main() {
    url := "amqp://guest:guest@localhost:5672"
    conn, err := amqp.Dial(url)
    if err != nil {
        log.Fatalf("connection.open: %s", err)
    }   

    wg := new(sync.WaitGroup)

    go func() {
        wg.Add(1)
        time.Sleep(time.Second)
        conn.Channel()
        wg.Done()
    }() 

    time.Sleep(time.Second)
    conn.Close()
    wg.Wait()
}
```

Our solution is to add a `closed` variable to the connection and reference it (always under lock) before allocating a channel. The code is https://github.com/2nd/amqp/commit/fb0a3d94dd9dde2bb44be49057d0adb512b4640c  (see commit message for more details).

We're not submitting this as a PR because we're just pushing it live today and don't feel confident pushing this on anyone else at this point. However, we'd welcome feedback on the issue & solution.

**Comment by karlseguin** [comment]
Follow up:
While the commit solved the above code, we saw the error again. A follow up commit is currently looking more promising: https://github.com/2nd/amqp/commit/0b08b93c2157c997656d1e44c0519f03eb4121a8  (however, individually, this change didn't solve the issue either, so either the issue is still present, or only both changes together resolve it).

Channel's `call` method appears to handle `rpc` being closed, but nowhere does it appear to actually get closed. Looking through past commits, closing `rpc` as a means of breaking that select appears to have been replaced with writing to `errors` (though much of that code was kept around). If the issue is in fact resolved, it would appear to be indicate that the channel's shutdown is being closed without receiving an error, which doesn't quite seem possible to me...but, if it works...

Tangentially, if all else fails, we'll probably timeout that select. I rather deal with eager exits than no exits.

**Comment by karlseguin** [comment]
Grokked.

The reason protecting the channels map, as seen in our first commit, didn't work is because of the following sequence:

Thread 1 - conn.Channnel()
Thread 1 - openChannel()
Thread 1 - allocateChannel()
Thread 1 - lock + insert in the map
Thread 1 - unlock, exit allocateChannel()
Thread X - connection dies
Thread X - connection.shutdown
Thread X - closes all channels, including the channel being created by Thread 1
Thread 1 - still in openChannel(), calls ch.open()
Thread 1 - deadlocked

In other words, it's possible for channel.call to be called despite having been closed. We could fix this by broadening the lock from `allocateChannel` to all of `openChannel`, but there might be other places where `ch.call` is called on a closed channel. Therefore, I feel comfortable that our 2nd commit does, in fact, solve this.

You can reproduce the issue by grabbing this branch which slows openChannel with a sleep:
https://github.com/2nd/amqp/commit/34e5fdeabd6909cf4f915412e3709c77795a9379

and running this code:

```
func main() {
	url := "amqp://guest:guest@localhost:5672"
	conn, err := amqp.Dial(url)
	if err != nil {
		panic(err)
	}
	wg := new(sync.WaitGroup)

	go func() {
		wg.Add(1)
		fmt.Println(conn.Channel())
		wg.Done()
	}()

	time.Sleep(time.Second)
	conn.Close()
	wg.Wait()
}
```

**Comment by karlseguin** [comment]
Just another update:

closing the rpc channel didn't seem to fix all the problems. Also, it's resulted in some code now trying to write to the closed channels. However, we've done other "fixes" to try to solve various other concurrency issues, so maybe it's related to one of those.

I can say that adding a timeout to that code *does* appear to have solve the issue. Something like:

```
select  {
case <-time.After(time.Second * 10):
  if me.connection.closed {
    return ErrClosed
  }
  me.Close()
  return ErrClosed
case e := <-me.errors:
    ...
case msg := <-me.rpc:
    ...
}
```

Note that this method isn't used when consuming from a queue...so it won't cause a timeout if you're just blocking while reading messages. 

Of course, given that we still don't have it working properly, you should take all of this with a grain of sand.

**Comment by lud4ik** [comment]
We have exact same issue. To what final solution have you come?

**Comment by karlseguin** [comment]
We have a fork:
https://github.com/2nd/amqp

Seems to be working, though we haven't pushed much load through it lately.

**Comment by lud4ik** [comment]
What is the reason of this change in your fork:
```
 	id, ok := me.allocator.next()
 	if !ok {
 		return nil, ErrChannelMax
 	}
-	ch := newChannel(me, uint16(id))
-	me.channels[uint16(id)] = ch
+	id16 := uint16(id)
+	ch := newChannel(me, id16)
+	if existing, exists := me.channels[id16]; exists {
+		existing.Close()
+	}
+	me.channels[id16] = ch
```
Did you have some issue with allocator?

**Comment by lud4ik** [comment]
What do you think of using closing errors channel and handle nil as ErrClosed. It fixes both cases, too.
It's important to handle the same for 0 channel, which is dispatched in connection.
Also I didn't touch your change to have entire lock on conn shutdown. It is because of me.closes and other conn fields could be mutated in other methods, isn't it? 
This approach introduce less locks (I mean without lock on every write to rpc channel):
```
diff --git a/channel.go b/channel.go
index 8976fa9..90e6d2b 100644
--- a/channel.go
+++ b/channel.go
@@ -106,6 +106,8 @@ func (me *Channel) shutdown(e *Error) {
 		// Notify RPC if we're selecting
 		if e != nil {
 			me.errors <- e
+		} else {
+			close(me.errors)
 		}
 
 		me.consumers.closeAll()
@@ -148,6 +150,9 @@ func (me *Channel) call(req message, res ...message) error {
 	if req.wait() {
 		select {
 		case e := <-me.errors:
+			if e == nil {
+				return ErrClosed
+			}
 			return e
 
 		case msg := <-me.rpc:
diff --git a/connection.go b/connection.go
index ad40079..9192de7 100644
--- a/connection.go
+++ b/connection.go
@@ -93,6 +93,7 @@ type Connection struct {
 	Major      int   // Server's major version
 	Minor      int   // Server's minor version
 	Properties Table // Server properties
+	closed     bool
 }
 
 type readDeadliner interface {
@@ -340,6 +341,10 @@ func (me *Connection) send(f frame) error {
 
 func (me *Connection) shutdown(err *Error) {
 	me.destructor.Do(func() {
+		me.m.Lock()
+		me.closed = true
+		me.m.Unlock()
+
 		if err != nil {
 			for _, c := range me.closes {
 				c <- err
@@ -352,6 +357,8 @@ func (me *Connection) shutdown(err *Error) {
 
 		if err != nil {
 			me.errors <- err
+		} else {
+			close(me.errors)
 		}
 
 		me.conn.Close()
@@ -545,6 +552,9 @@ func (me *Connection) allocateChannel() (*Channel, error) {
 	}
 
 	ch := newChannel(me, uint16(id))
+	if me.closed {
+		return nil, ErrClosed
+	}
 	me.channels[uint16(id)] = ch
 
 	return ch, nil
@@ -602,6 +612,9 @@ func (me *Connection) call(req message, res ...message) error {
 
 	select {
 	case err := <-me.errors:
+		if err == nil {
+			return ErrClosed
+		}
 		return err
 
 	case msg := <-me.rpc:

```

**Comment by karlseguin** [comment]
We didn't catch any allocator issues, it was a suspicion we were having. To be honest, there are still a few things I'm not sure *how* they are happening.

The conn.shutdown lock is, as far as I can tell, absolutely essential (though not directly related to this deadlock issue). I think there's more than 1 issue there, but, as an example, it's possible to iterate through me.channels, closing the channels and then have another thread call `allocateChannel`.

Your solution looks reasonable.

**Comment by lud4ik** [comment]
It didn't solve all issues. I just don't get how it can happen. I added more logging, but it doesn't reproduce with it, I removed some of them and we have situation with "second channel  open seen" and it's blocked on same select with invoked connection shutdown. I've read all connection and channel code, but don't know what else I can do...

**Comment by michaelklishin** [comment]
Addressed in #188.
